File: src/main/java/net/gamershub/kitpvp/abilities/AbilityHandler.java
package net.gamershub.kitpvp.abilities;

import net.gamershub.kitpvp.KitPvpPlugin;
import net.gamershub.kitpvp.StringArrayPersistentDataType;
import net.gamershub.kitpvp.Utils;
import net.gamershub.kitpvp.abilities.impl.FireballAbility;
import net.gamershub.kitpvp.abilities.impl.LightningAbility;
import net.gamershub.kitpvp.abilities.impl.ThunderstormAbility;
import org.bukkit.NamespacedKey;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.inventory.ItemStack;
import org.bukkit.persistence.PersistentDataContainer;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

public class AbilityHandler implements Listener {
    public HashMap<String, Ability> abilities = new HashMap<>();

    public static Ability FIREBALL;
    public static Ability LIGHTNING;
    public static Ability THUNDERSTORM;

    public AbilityHandler() {
        FIREBALL = registerAbility(new FireballAbility());
        LIGHTNING = registerAbility(new LightningAbility());
        THUNDERSTORM = registerAbility(new ThunderstormAbility());
    }

    public Ability registerAbility(Ability ability) {
        Utils.registerEvents(ability);

        if (!abilities.containsKey(ability.getId())) {
            abilities.put(ability.getId(), ability);
        } else {
            throw new IllegalArgumentException("Ability ID already in use");
        }

        return ability;
    }

    public List<Ability> getItemAbilities(ItemStack itemStack) {
        PersistentDataContainer container = itemStack.getItemMeta().getPersistentDataContainer();
        NamespacedKey key = new NamespacedKey(KitPvpPlugin.INSTANCE, "abilities");

        if (container.has(key)) {
            String[] abilities = container.get(key, new StringArrayPersistentDataType());
            if (abilities == null) return Collections.emptyList();

            List<Ability> abilityList = new ArrayList<>();

            for (String id : abilities) {
                abilityList.add(this.abilities.get(id));
            }

            return abilityList;

        }

        return Collections.emptyList();
    }

    @EventHandler
    public void onInteract(PlayerInteractEvent e) {
        if (e.getItem() == null) return;

        PersistentDataContainer container = e.getItem().getItemMeta().getPersistentDataContainer();
        NamespacedKey key = new NamespacedKey(KitPvpPlugin.INSTANCE, "abilities");

        if (container.has(key)) {
            String[] abilities = container.get(key, new StringArrayPersistentDataType());
            if (abilities == null) return;
            for (String id : abilities) {
                if (this.abilities.containsKey(id)) {
                    Ability ability = this.abilities.get(id);
                    Player p = e.getPlayer();

                    switch (ability.getType()) {
                        case RIGHT_CLICK -> {
                            if (e.getAction().isRightClick()) ability.handleAbility(e);
                        }
                        case LEFT_CLICK -> {
                            if (e.getAction().isLeftClick()) ability.handleAbility(e);
                        }
                        case SNEAK_RIGHT_CLICK -> {
                            if (e.getAction().isRightClick() && p.isSneaking()) ability.handleAbility(e);
                        }
                        case SNEAK_LEFT_CLICK -> {
                            if (e.getAction().isLeftClick() && p.isSneaking()) ability.handleAbility(e);
                        }
                    }
                }
            }
        }
    }
}

File: src/main/kotlin/io/github/chaosdave34/kitpvp/abilities/AbilityHandler.kt
package io.github.chaosdave34.kitpvp.abilities

import io.github.chaosdave34.ghutils.Utils
import io.github.chaosdave34.ghutils.persistentdatatypes.StringArrayPersistentDataType
import io.github.chaosdave34.kitpvp.ExtendedPlayer
import io.github.chaosdave34.kitpvp.KitPvp
import io.github.chaosdave34.kitpvp.abilities.impl.archer.LeapAbility
import io.github.chaosdave34.kitpvp.abilities.impl.artilleryman.EnhanceAbility
import io.github.chaosdave34.kitpvp.abilities.impl.assassin.HauntAbility
import io.github.chaosdave34.kitpvp.abilities.impl.creeper.ChargeAbility
import io.github.chaosdave34.kitpvp.abilities.impl.creeper.ExplodeAbility
import io.github.chaosdave34.kitpvp.abilities.impl.creeper.FireballAbility
import io.github.chaosdave34.kitpvp.abilities.impl.devil.FireStormAbility
import io.github.chaosdave34.kitpvp.abilities.impl.enderman.DragonFireballAbility
import io.github.chaosdave34.kitpvp.abilities.impl.enderman.EnderAttackAbility
import io.github.chaosdave34.kitpvp.abilities.impl.engineer.ModularShieldAbility
import io.github.chaosdave34.kitpvp.abilities.impl.engineer.OverloadAbility
import io.github.chaosdave34.kitpvp.abilities.impl.engineer.TurretAbility
import io.github.chaosdave34.kitpvp.abilities.impl.magician.LevitateAbility
import io.github.chaosdave34.kitpvp.abilities.impl.magician.ShuffleAbility
import io.github.chaosdave34.kitpvp.abilities.impl.poseidon.StormAbility
import io.github.chaosdave34.kitpvp.abilities.impl.poseidon.WaterBurstAbility
import io.github.chaosdave34.kitpvp.abilities.impl.spacesoldier.AirstrikeAbility
import io.github.chaosdave34.kitpvp.abilities.impl.spacesoldier.BlackHoleAbility
import io.github.chaosdave34.kitpvp.abilities.impl.spacesoldier.DarknessAbility
import io.github.chaosdave34.kitpvp.abilities.impl.tank.GroundSlamAbility
import io.github.chaosdave34.kitpvp.abilities.impl.tank.TornadoAbility
import io.github.chaosdave34.kitpvp.abilities.impl.vampire.BatMorhpAbility
import io.github.chaosdave34.kitpvp.abilities.impl.zeus.LightningAbility
import io.github.chaosdave34.kitpvp.abilities.impl.zeus.RageAbility
import io.github.chaosdave34.kitpvp.abilities.impl.zeus.ThunderstormAbility
import org.bukkit.NamespacedKey
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.event.player.PlayerToggleSneakEvent
import org.bukkit.inventory.ItemStack

class AbilityHandler : Listener {
    val abilities: MutableMap<String, Ability> = mutableMapOf()

    companion object {
        @JvmStatic
        lateinit var FIREBALL: Ability

        @JvmStatic
        lateinit var LIGHTNING: Ability

        @JvmStatic
        lateinit var THUNDERSTORM: Ability

        @JvmStatic
        lateinit var HAUNT: Ability

        @JvmStatic
        lateinit var LEVITATE: Ability

        @JvmStatic
        lateinit var SHUFFLE: Ability

        @JvmStatic
        lateinit var BAT_MORPH: Ability

        @JvmStatic
        lateinit var AIRSTRIKE: Ability

        @JvmStatic
        lateinit var DARKNESS: Ability

        @JvmStatic
        lateinit var EXPLODE: Ability

        @JvmStatic
        lateinit var ENDER_ATTACK: Ability

        @JvmStatic
        lateinit var STORM: Ability

        @JvmStatic
        lateinit var GROUND_SLAM: Ability

        @JvmStatic
        lateinit var TORNADO: Ability

        @JvmStatic
        lateinit var ENHANCE: Ability

        @JvmStatic
        lateinit var TURRET: Ability

        @JvmStatic
        lateinit var LEAP: Ability

        @JvmStatic
        lateinit var DRAGON_FIREBALL: Ability

        @JvmStatic
        lateinit var FIRE_STORM: Ability

        @JvmStatic
        lateinit var RAGE: Ability

        @JvmStatic
        lateinit var MODULAR_SHIELD: Ability

        @JvmStatic
        lateinit var OVERLOAD: Ability

        @JvmStatic
        lateinit var CHARGE: Ability

        @JvmStatic
        lateinit var WATER_BURST: Ability

        @JvmStatic
        lateinit var BLACK_HOLE: Ability
    }

    init {
        FIREBALL = registerAbility(FireballAbility())
        LIGHTNING = registerAbility(LightningAbility())
        THUNDERSTORM = registerAbility(ThunderstormAbility())
        HAUNT = registerAbility(HauntAbility())
        LEVITATE = registerAbility(LevitateAbility())
        SHUFFLE = registerAbility(ShuffleAbility())
        BAT_MORPH = registerAbility(BatMorhpAbility())
        AIRSTRIKE = registerAbility(AirstrikeAbility())
        DARKNESS = registerAbility(DarknessAbility())
        EXPLODE = registerAbility(ExplodeAbility())
        ENDER_ATTACK = registerAbility(EnderAttackAbility())
        STORM = registerAbility(StormAbility())
        GROUND_SLAM = registerAbility(GroundSlamAbility())
        TORNADO = registerAbility(TornadoAbility())
        ENHANCE = registerAbility(EnhanceAbility())
        TURRET = registerAbility(TurretAbility())
        LEAP = registerAbility(LeapAbility())
        DRAGON_FIREBALL = registerAbility(DragonFireballAbility())
        FIRE_STORM = registerAbility(FireStormAbility())
        RAGE = registerAbility(RageAbility())
        MODULAR_SHIELD = registerAbility(ModularShieldAbility())
        OVERLOAD = registerAbility(OverloadAbility())
        CHARGE = registerAbility(ChargeAbility())
        WATER_BURST = registerAbility(WaterBurstAbility())
        BLACK_HOLE = registerAbility(BlackHoleAbility())
    }

    private fun registerAbility(ability: Ability): Ability {
        Utils.registerEvents(ability)
        abilities[ability.id] = ability
        return ability
    }

    fun getItemAbilities(itemStack: ItemStack): List<Ability> {
        val container = itemStack.itemMeta.persistentDataContainer
        val key = NamespacedKey(KitPvp.INSTANCE, "abilities")

        if (container.has(key)) {
            val abilities = container.get(key, StringArrayPersistentDataType()) ?: return emptyList()

            val abilityList: MutableList<Ability> = ArrayList()

            for (id in abilities) {
                this.abilities[id]?.let { abilityList.add(it) }
            }
            return abilityList
        }
        return emptyList()
    }

    @EventHandler
    fun onInteract(event: PlayerInteractEvent) {
        val player = event.player
        val item = event.item ?: return

        if (ExtendedPlayer.from(player).inSpawn()) return

        val container = item.itemMeta.persistentDataContainer
        val key = NamespacedKey(KitPvp.INSTANCE, "abilities")

        if (container.has(key)) {
            val abilities = getItemAbilities(item)
            val abilityTypes = abilities.stream().map(Ability::type).toList()

            for (ability in abilities) {
                when (ability.type) {
                    Ability.Type.RIGHT_CLICK -> {
                        if (event.action.isRightClick) {
                            if (abilityTypes.contains(Ability.Type.SNEAK_RIGHT_CLICK) && player.isSneaking) continue
                            ability.handleAbility(player)
                        }
                    }

                    Ability.Type.LEFT_CLICK -> {
                        if (event.action.isLeftClick) {
                            if (abilityTypes.contains(Ability.Type.SNEAK_LEFT_CLICK) && player.isSneaking) continue
                            ability.handleAbility(player)
                        }
                    }

                    Ability.Type.SNEAK_RIGHT_CLICK -> {
                        if (event.action.isRightClick && player.isSneaking) ability.handleAbility(player)
                    }

                    Ability.Type.SNEAK_LEFT_CLICK -> {
                        if (event.action.isLeftClick && player.isSneaking) ability.handleAbility(player)
                    }

                    else -> continue
                }
            }
        }
    }

    @EventHandler
    fun onSneak(event: PlayerToggleSneakEvent) {
        val player = event.player
        if (!event.isSneaking) return
        if (ExtendedPlayer.from(player).inSpawn()) return

        for (armorContent in player.inventory.armorContents) {
            if (armorContent == null) continue

            val container = armorContent.itemMeta.persistentDataContainer
            val key = NamespacedKey(KitPvp.INSTANCE, "abilities")
            if (container.has(key)) {
                val abilities = container.get(key, StringArrayPersistentDataType()) ?: return
                for (id in abilities) {
                    if (this.abilities.containsKey(id)) {
                        val ability = this.abilities[id]

                        if (ability?.type == Ability.Type.SNEAK) ability.handleAbility(player)
                    }
                }
            }
        }
    }
}